File: backend_python/app/repositories/refresh_token.py
"""Refresh token repository for token storage operations."""

import hashlib
from datetime import datetime, timezone

from sqlmodel import Session, select

from app.models import RefreshToken
from app.repositories.base import BaseRepository


class RefreshTokenRepository(BaseRepository[RefreshToken]):
    """Repository for RefreshToken model operations."""

    def __init__(self, session: Session):
        super().__init__(session, RefreshToken)

    @staticmethod
    def hash_token(token: str) -> str:
        """Hash a token for secure storage."""
        return hashlib.sha256(token.encode()).hexdigest()

    def create(self, token: str, user_id: int, expires_at: datetime) -> RefreshToken:
        """Create and store a new refresh token.

        Args:
            token: The raw refresh token string
            user_id: The user's database ID
            expires_at: When the token expires

        Returns:
            The created RefreshToken record
        """
        refresh_token = RefreshToken(
            token_hash=self.hash_token(token),
            user_id=user_id,
            expires_at=expires_at,
            revoked=False,
        )
        self.session.add(refresh_token)
        self.session.commit()
        self.session.refresh(refresh_token)
        return refresh_token

    def get_by_token(self, token: str) -> RefreshToken | None:
        """Find a refresh token by its raw value.

        Args:
            token: The raw refresh token string

        Returns:
            The RefreshToken record if found and valid, None otherwise
        """
        token_hash = self.hash_token(token)
        statement = select(RefreshToken).where(
            RefreshToken.token_hash == token_hash,
            RefreshToken.revoked == False,  # noqa: E712
            RefreshToken.expires_at > datetime.now(timezone.utc),
        )
        return self.session.exec(statement).first()

    def revoke(self, token: str) -> bool:
        """Revoke a refresh token.

        Args:
            token: The raw refresh token string

        Returns:
            True if token was found and revoked, False otherwise
        """
        token_hash = self.hash_token(token)
        statement = select(RefreshToken).where(RefreshToken.token_hash == token_hash)
        refresh_token = self.session.exec(statement).first()

        if refresh_token:
            refresh_token.revoked = True
            self.session.add(refresh_token)
            self.session.commit()
            return True
        return False

    def revoke_all_for_user(self, user_id: int) -> int:
        """Revoke all refresh tokens for a user.

        Args:
            user_id: The user's database ID

        Returns:
            Number of tokens revoked
        """
        statement = select(RefreshToken).where(
            RefreshToken.user_id == user_id,
            RefreshToken.revoked == False,  # noqa: E712
        )
        tokens = self.session.exec(statement).all()

        for token in tokens:
            token.revoked = True
            self.session.add(token)

        self.session.commit()
        return len(tokens)

    def cleanup_expired(self) -> int:
        """Delete expired and revoked tokens.

        Returns:
            Number of tokens deleted
        """
        statement = select(RefreshToken).where(
            (RefreshToken.expires_at < datetime.now(timezone.utc))
            | (RefreshToken.revoked == True)  # noqa: E712
        )
        tokens = self.session.exec(statement).all()

        for token in tokens:
            self.session.delete(token)

        self.session.commit()
        return len(tokens)

File: backend/app/models/refresh_token.rb
require 'digest'
require 'securerandom'

# Opaque refresh tokens for the JWT auth gateway (B2).
#
# Parity with Ronald's FastAPI implementation
# (rdom_test_new_sushi_20251030/backend_python/app/repositories/refresh_token.py):
# the raw token is a cryptographically random string handed to the client in an
# HttpOnly cookie; only its SHA-256 hash is persisted. B2 additionally performs
# rotation + reuse detection in Api::V1::AuthController#refresh.
class RefreshToken < ApplicationRecord
  belongs_to :user

  # Active = not revoked and not expired.
  scope :active, -> { where(revoked: false).where('expires_at > ?', Time.current) }

  # Is this table actually present in the database we are connected to?
  #
  # 082's database IS legacy production SUSHI's, which has no `refresh_tokens`
  # table and must never gain one — no migration runs on that node, by standing
  # decision. Without this check a successful login there raises
  # ActiveRecord::StatementInvalid on the INSERT and answers 500, so LDAP could
  # not be exercised on the production node at all.
  #
  # Memoized per process: a login must not cost a schema query, and the answer
  # cannot change while the process lives. Rails' reloader resets the class in
  # development, which re-asks — that is the behaviour we want after a migration.
  def self.available?
    return @available unless @available.nil?

    @available = connection.table_exists?(table_name)
  rescue StandardError => e
    Rails.logger.error("RefreshToken.available? check failed: #{e.message}")
    @available = false
  end

  # SHA-256 hex digest used as the at-rest representation of a raw token.
  def self.hash_token(raw)
    Digest::SHA256.hexdigest(raw.to_s)
  end

  # Issue a new refresh token for +user+, valid for +ttl+.
  # Returns [record, raw_token]; the raw token is only available here (never stored).
  def self.issue(user:, ttl:)
    raw = SecureRandom.urlsafe_base64(32)
    record = create!(
      user: user,
      token_hash: hash_token(raw),
      expires_at: ttl.from_now,
      revoked: false
    )
    [record, raw]
  end

  # Lookup a valid (non-revoked, non-expired) token by its raw value.
  def self.find_active(raw)
    return nil if raw.blank?

    active.find_by(token_hash: hash_token(raw))
  end

  # Lookup by raw value regardless of revoked/expired state (for reuse detection).
  def self.find_any(raw)
    return nil if raw.blank?

    find_by(token_hash: hash_token(raw))
  end

  # Revoke a single token by its raw value. Returns true if a token was revoked.
  def self.revoke!(raw)
    record = find_any(raw)
    return false unless record

    record.update!(revoked: true) unless record.revoked?
    true
  end

  # Revoke every active token for +user+. Returns the number revoked.
  def self.revoke_all_for!(user)
    active.where(user_id: user.id).update_all(revoked: true, updated_at: Time.current)
  end
end
